src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,357 @@
+from flint.flintlib.fmpz cimport (
+    fmpz_t,
+    fmpz_one,
+    fmpz_set,
+    fmpz_init,
+    fmpz_clear,
+    fmpz_equal
+)
+from flint.flintlib.fmpz_mod cimport *
+
+from flint.utils.typecheck cimport typecheck
+from flint.flint_base.flint_base cimport flint_scalar
+from flint.types.fmpz cimport (
+    fmpz,
+    any_as_fmpz,
+    fmpz_get_intlong
+)
+
+
+cdef class fmpz_mod_ctx:
+    """
+    Context object for *fmpz_mod* initalised with a modulus `n`
+
+        >>> fmpz_mod_ctx(2**127 - 1)
+        fmpz_mod_ctx(170141183460469231731687303715884105727)
+
+    """
+    def __cinit__(self):
+        # TODO: is this the best method?
+        cdef fmpz one = fmpz.__new__(fmpz)
+        fmpz_one(one.val)
+        fmpz_mod_ctx_init(self.val, one.val)
+
+    def __dealloc__(self):
+        fmpz_mod_ctx_clear(self.val)
+
+    def __init__(self, mod):
+        # Ensure modulus is fmpz type
+        if not typecheck(mod, fmpz):
+            mod = any_as_fmpz(mod)
+            if mod is NotImplemented:
+                raise TypeError("Context modulus must be able to be case to an `fmpz` type")
+
+        # Ensure modulus is positive
+        if mod < 1:
+            raise ValueError("Modulus is expected to be positive")
+
+        # Init the context
+        fmpz_mod_ctx_init(self.val, (<fmpz>mod).val)
+    
+    def modulus(self):
+        """
+        Return the modulus from the context as an fmpz
+        type
+
+            >>> mod_ctx = fmpz_mod_ctx(2**127 - 1)
+            >>> mod_ctx.modulus()
+            170141183460469231731687303715884105727
+
+        """
+        n = fmpz()
+        fmpz_set(n.val, (<fmpz_t>self.val.n))
+        return n
+
+    def __eq__(self, other):
+        if typecheck(other, fmpz_mod_ctx):
+            return fmpz_equal(self.val.n, (<fmpz_mod_ctx>other).val.n)
+        return False
+
+    def __hash__(self):
+        return hash(self.modulus())
+
+    def __str__(self):
+        return f"Context for fmpz_mod with modulus: {self.modulus()}"
+
+    def __repr__(self):
+        return f"fmpz_mod_ctx({self.modulus()})"
+
+    def __call__(self, val):
+        return fmpz_mod(val, self)
+
+cdef class fmpz_mod(flint_scalar):
+    """
+    The *fmpz_mod* type represents integer modulo an 
+    arbitrary-size modulus. For wordsize modulus, see
+    *nmod*.
```

Comment:
A Sphinx reference like
```
:class:`flint.types.nmod.nmod`
```
creates a clickable link in the html docs which is better.

Reply:
Ahh thanks, I haven't ever used sphinx. I'll add this now.

Reply:
Actually this is a shorter version:
```
:class:`~.nmod`
```